gumpy/features.py

```diff
@@ -249,38 +249,7 @@ def PCA_dim_red(features, var_desired):
             return features_reduced

```

Comment:
Overall, the code seems quite clean and well-structured. However, there are a few improvements that could be made:

1. **Docstrings**: The docstrings of some methods are not complete. For example, `CSP` method's docstring does not explain what is expected as input. Similarly, `spatialFilter` and `covarianceMatrix` methods' docstrings do not provide details about the input parameters. It would be beneficial to provide a more detailed explanation of the parameters and expected input format.

2. **Error Handling**: In the `sequential_feature_selector` method, if the `selection_type` is not one of the predefined types, the function raises a generic Exception. It would be better to define a custom exception for this case, which would make it easier to catch and handle this specific error elsewhere in the code.

3. **Code Duplication**: In the `sequential_feature_selector` method, the code to create `sfs` is repeated with only one parameter changing. This could be simplified by creating a dictionary mapping selection types to their parameters, and then using this dictionary to create `sfs`.

4. **Variable Names**: Some variable names are not descriptive enough (e.g., `w`, `n` in `dwt_features` method). It's always a good practice to use descriptive variable names to make the code more readable and maintainable.

5. **Magic Numbers**: The code contains magic numbers, like `4` in `powermean` method. It's recommended to define such numbers as constants at the beginning of the script, giving them descriptive names.

6. **Commenting**: Some parts of the code lack comments, making it harder to understand what the code is doing. Adding comments to explain the logic would make the code easier to maintain and understand. 

7. **Imports**: The import statements are scattered throughout the code. It's a good practice to keep all the import statements at the beginning of the file.

Here is a sample of how some of these changes could be implemented:

```python
# Define constants
CONSTANT_4 = 4

# Define custom exception
class UnknownSelectionTypeError(Exception):
    pass

# Define selection types
SELECTION_TYPES = {
    'SFS': {"forward": True, "floating": False},
    'SBS': {"forward": False, "floating": False},
    'SFFS': {"forward": True, "floating": True},
    'SBFS': {"forward": True, "floating": True},
}

def sequential_feature_selector(features, labels, classifier, k_features, kfold, selection_type, plot=True, **kwargs):
    ...
    # Check if selection_type is valid
    if selection_type not in SELECTION_TYPES:
        raise UnknownSelectionTypeError(f"Unknown selection type '{selection_type}'")

    sfs = SFS(clf, k_features, verbose=2, scoring='accuracy', cv=kfold, n_jobs=-1, **SELECTION_TYPES[selection_type])
    ...

def powermean(data, trial, fs, window):
    return np.power(data[trial+fs*CONSTANT_4+window[0]: trial+fs*CONSTANT_4+window[1],0],2).mean(), \
           np.power(data[trial+fs*CONSTANT_4+window[0]: trial+fs*CONSTANT_4+window[1],1],2).mean(), \
           np.power(data[trial+fs*CONSTANT_4+window[0]: trial+fs*CONSTANT_4+window[1],2],2).mean()
``

Reply:
Hey, thanks for your very thoughtful comment! I agree with everything that you wrote. It hasn't reached the maturity I would love. However, the project is currently stale and we don't actively support it. If you open a reasonable pull request, I'm very inclined to have a look at it and potentially merge it :)